#중복 데이터
##1. 데이터 불러오기

In [1]:
import pandas as pd


In [2]:
df = pd.read_csv('data/bicycle_dup.csv', engine='python', encoding='cp949')
df

자전거번호              대여일시  대여소번호               대여소명  대여거치대  \
0    SPB-23220   2019-11-01 8:48    646  장한평역 1번출구 (국민은행앞)    3.0   
1    SPB-16216   2019-11-04 8:38    646  장한평역 1번출구 (국민은행앞)    2.0   
2    SPB-21097   2019-11-04 8:46    646  장한평역 1번출구 (국민은행앞)    1.0   
3    SPB-22292   2019-11-05 8:34    646  장한평역 1번출구 (국민은행앞)    1.0   
4    SPB-07935  2019-11-05 12:29    512        뚝섬역 1번 출구 옆   11.0   
..         ...               ...    ...                ...    ...   
415  SPB-12116   2019-11-15 4:22    112            극동방송국 앞    5.0   
416  SPB-17768   2019-11-15 9:31    103         망원역 2번출구 앞   14.0   
417  SPB-18841  2019-11-15 20:40    113       홍대입구역 2번출구 앞   24.0   
418  SPB-24467  2019-11-15 21:55    152         마포구민체육센터 앞   18.0   
419  SPB-17570  2019-11-15 22:16    391              정동길입구    7.0   

                 반납일시  반납대여소번호       반납대여소명  반납거치대  이용시간    이용거리   나이  
0     2019-11-01 9:01        3         중랑센터    7.0  12.0  1100.0   50  
1     2019-11-04 8:56        3         중랑센터    2.0   7.0  1100.0   50  
2     2019-11-04 8:57        3         중랑센터    7.0  10.0     NaN   32  
3     2019-11-05 8:45        3         중랑센터    NaN  10.0  1380.0   21  
4    2019-11-05 12:39        3         중랑센터    7.0  10.0  1650.0  120  
..                ...      ...          ...    ...   ...     ...  ...  
415   2019-11-15 4:33      101  (구)합정동 주민센터    1.0  11.0  2200.0   22  
416   2019-11-15 9:37      101  (구)합정동 주민센터    NaN   5.0     NaN   22  
417  2019-11-15 20:53      101  (구)합정동 주민센터    1.0  12.0  2280.0   37  
418  2019-11-15 22:05      101  (구)합정동 주민센터    1.0  10.0  1180.0   27  
419  2019-11-15 23:04      101  (구)합정동 주민센터    5.0  48.0  8100.0   28  

[420 rows x 12 columns]

In [3]:
df[df['이용시간'].isnull()]

자전거번호              대여일시  대여소번호               대여소명  대여거치대  \
11   SPB-20145   2019-11-13 8:30    646  장한평역 1번출구 (국민은행앞)    9.0   
16   SPB-22311   2019-11-18 8:32    646  장한평역 1번출구 (국민은행앞)    6.0   
22   SPB-05857   2019-11-26 8:40    646  장한평역 1번출구 (국민은행앞)    6.0   
29   SPB-21611   2019-11-27 8:34    911    은평평화공원(역촌역4번출구)    NaN   
44   SPB-14142  2019-11-01 13:06    152         마포구민체육센터 앞   27.0   
57   SPB-07887  2019-11-01 20:22    104         합정역 1번출구 앞    6.0   
76   SPB-07402  2019-11-02 16:45    151           망원1동주민센터    1.0   
86   SPB-04854  2019-11-02 19:52    101        (구)합정동 주민센터    5.0   
129  SPB-22403  2019-11-03 23:45    119            서강나루 공원   10.0   
161  SPB-04641  2019-11-04 21:33    101        (구)합정동 주민센터    3.0   
185  SPB-20864  2019-11-05 13:18    111         상수역 2번출구 앞   10.0   
228  SPB-17074  2019-11-06 18:33    267          삼성화재 사옥 옆    2.0   
311  SPB-18301  2019-11-09 23:40    103         망원역 2번출구 앞    9.0   
332  SPB-06200  2019-11-11 14:26    115          사루비아 빌딩 앞    1.0   
400  SPB-15391  2019-11-14 17:14    113       홍대입구역 2번출구 앞   22.0   

                 반납일시  반납대여소번호       반납대여소명  반납거치대  이용시간    이용거리   나이  
11    2019-11-13 8:38        3         중랑센터    2.0   NaN  1300.0   30  
16    2019-11-18 8:39        3         중랑센터    2.0   NaN  1350.0   31  
22    2019-11-26 8:48        3         중랑센터    2.0   NaN  1380.0   41  
29    2019-11-27 8:55        5     상암센터 정비실    6.0   NaN  5350.0   38  
44   2019-11-01 13:12      101  (구)합정동 주민센터    5.0   NaN  1190.0   40  
57   2019-11-01 20:30      101  (구)합정동 주민센터    3.0   NaN   770.0   31  
76   2019-11-02 16:59      101  (구)합정동 주민센터    3.0   NaN  1250.0   41  
86   2019-11-02 21:01      101  (구)합정동 주민센터    2.0   NaN  2300.0  220  
129  2019-11-03 23:59      101  (구)합정동 주민센터    5.0   NaN  1530.0   62  
161  2019-11-04 21:42      101  (구)합정동 주민센터    1.0   NaN     0.0   45  
185  2019-11-05 13:36      101  (구)합정동 주민센터    1.0   NaN  3680.0   41  
228  2019-11-06 18:49      101  (구)합정동 주민센터    2.0   NaN  2390.0   31  
311   2019-11-10 0:21      101  (구)합정동 주민센터    4.0   NaN  6510.0   22  
332  2019-11-11 14:44      101  (구)합정동 주민센터    4.0   NaN     0.0   41  
400  2019-11-14 17:29      101  (구)합정동 주민센터    1.0   NaN  2290.0   31

In [4]:
df[df['이용거리'] == 2290]

자전거번호              대여일시  대여소번호          대여소명  대여거치대  \
98   SPB-11651  2019-11-02 23:54    228   선유도역 3번출구 앞    8.0   
400  SPB-15391  2019-11-14 17:14    113  홍대입구역 2번출구 앞   22.0   

                 반납일시  반납대여소번호       반납대여소명  반납거치대  이용시간    이용거리  나이  
98    2019-11-03 0:09      101  (구)합정동 주민센터    3.0  15.0  2290.0  23  
400  2019-11-14 17:29      101  (구)합정동 주민센터    1.0   NaN  2290.0  31

In [5]:
df_group = df.groupby('이용거리')
df_group.mean(numeric_only=True)

대여소번호     대여거치대  반납대여소번호     반납거치대   이용시간          나이
이용거리                                                                
0.0       130.583333  8.583333    101.0  2.272727   15.8   29.416667
20.0      101.000000  2.000000    101.0  3.000000    2.0   48.000000
280.0     260.000000  2.000000     53.0  3.500000    4.0  278.000000
440.0     106.000000  2.000000    101.0  1.000000   12.0   57.000000
450.0     262.500000  8.000000     53.0  4.000000   23.5   32.500000
...              ...       ...      ...       ...    ...         ...
10260.0  1160.000000  1.000000    101.0  1.000000   47.0   25.000000
12610.0  1925.000000  6.000000    101.0  4.000000   62.0   66.000000
16910.0  2103.000000  1.000000    101.0  1.000000  121.0   46.000000
17060.0  2103.000000  3.000000    101.0  4.000000  121.0   24.000000
42070.0   116.000000  3.000000    101.0  4.000000   46.0   22.000000

[256 rows x 6 columns]

In [6]:
df_group.mean(numeric_only=True).loc[42070.0]

대여소번호      116.0
대여거치대        3.0
반납대여소번호    101.0
반납거치대        4.0
이용시간        46.0
나이          22.0
Name: 42070.0, dtype: float64

##2. 데이터 중복 확인
> DataFrame에서는 중복 확인을 위해 duplicated 함수를 지원한다. 중복된 데이터가 나타나면 true로 변환시킨다. 

In [7]:
df.duplicated(['이용거리'])

0      False
1       True
2      False
3      False
4      False
       ...  
415    False
416     True
417     True
418     True
419    False
Length: 420, dtype: bool

In [8]:
df.duplicated(['이용거리', '나이'])

0      False
1       True
2      False
3      False
4      False
       ...  
415    False
416     True
417    False
418    False
419    False
Length: 420, dtype: bool

>0번과 1번의 이용거리가 같아 1번 이용거리가 True로 설정되었다. 416~418까지도 중복된 데이터이기때문에 True로 설정되었다.

>두 가지 이상의 옵션을 설정하게 되면 and로 중복을 확인할 수 있다.

In [9]:
df.duplicated(['이용거리', '나이'], keep='last')

0       True
1      False
2      False
3      False
4      False
       ...  
415    False
416    False
417    False
418    False
419    False
Length: 420, dtype: bool

In [11]:
df.duplicated(['이용거리'], keep=False)

0       True
1       True
2       True
3       True
4      False
       ...  
415    False
416     True
417     True
418     True
419    False
Length: 420, dtype: bool

In [13]:
df.duplicated(['이용거리'], keep=False).sum()

np.int64(244)

In [14]:
df.drop_duplicates(['이용거리'], keep=False)

자전거번호              대여일시  대여소번호               대여소명  대여거치대  \
4    SPB-07935  2019-11-05 12:29    512        뚝섬역 1번 출구 옆   11.0   
15   SPB-13973   2019-11-15 8:42    646  장한평역 1번출구 (국민은행앞)    1.0   
26   SPB-15211   2019-11-15 8:42    419             홈플러스 앞   18.0   
27   SPB-24468   2019-11-22 8:36    911    은평평화공원(역촌역4번출구)   10.0   
28   SPB-09287   2019-11-26 8:36    911    은평평화공원(역촌역4번출구)    4.0   
..         ...               ...    ...                ...    ...   
410  SPB-18614  2019-11-14 23:07    391              정동길입구    5.0   
413  SPB-25161   2019-11-15 3:13    106         합정역 7번출구 앞    6.0   
414  SPB-24515   2019-11-15 3:58    112            극동방송국 앞    1.0   
415  SPB-12116   2019-11-15 4:22    112            극동방송국 앞    5.0   
419  SPB-17570  2019-11-15 22:16    391              정동길입구    7.0   

                 반납일시  반납대여소번호       반납대여소명  반납거치대  이용시간    이용거리   나이  
4    2019-11-05 12:39        3         중랑센터    7.0  10.0  1650.0  120  
15    2019-11-15 8:50        3         중랑센터    2.0   6.0  1370.0   32  
26    2019-11-15 9:16        5     상암센터 정비실    5.0  33.0   550.0   43  
27    2019-11-22 8:58        5     상암센터 정비실    6.0  20.0  5170.0   22  
28    2019-11-26 9:21        5     상암센터 정비실    6.0  42.0  5800.0   57  
..                ...      ...          ...    ...   ...     ...  ...  
410  2019-11-14 23:49      101  (구)합정동 주민센터    3.0  40.0  7510.0   28  
413   2019-11-15 3:18      101  (구)합정동 주민센터    1.0   3.0   640.0   21  
414   2019-11-15 4:08      101  (구)합정동 주민센터    1.0   9.0  1620.0   21  
415   2019-11-15 4:33      101  (구)합정동 주민센터    1.0  11.0  2200.0   22  
419  2019-11-15 23:04      101  (구)합정동 주민센터    5.0  48.0  8100.0   28  

[176 rows x 12 columns]

>keep 옵션은 기본값이 first이다. first는 첫번째 값을 기준으로 중복된 데이터가 나오면 true로 변환시키는 것이고 last는 중복된 가장 마지막 값을 false로 지정하는 것이다.

keep이 false이면 중복된 데이터를 모두 true로 변환시킨다.

##3. 중복 데이터 제거
>중복된 데이터를 제거하기 위해서는 drop_duplicates() 함수를 이용하면 된다. 사용방법은 중복데이터 확인과 같다.